### PR TITLE
copy expvalcost code for qnode compatibility in ShotAdaptiveOptimizer

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
We deprecated `qml.ExpvalCost` in v0.24, but it is the only way to use the multinomial distribution for distributing shots with the `ShotAdaptiveOptimizer`. Since we recommend simply using a QNode with the expectation of a Hamiltonian, this method should have feature parity with ExpvalCost.

**Description of the Change:**
Copied the multinomial distribution code with tweaks in order to support its use with regular (non-ExpvalCost) qnodes.

**Benefits:**
We are closer to removing ExpvalCost

**Possible Drawbacks:**
- Duplicate code for the time-being
- some not-so-pretty code pieces, like changing `QNode.func` in place.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -217,6 +217,11 @@
 * `Conditional` and `MeasurementValue` objects now implement `map_wires`.
   [(#4884)](https://github.com/PennyLaneAI/pennylane/pull/4884)
 
+* `qml.ShotAdaptiveOptimizer` can now use a multinomial distribution for spreading shots across
+  the terms of a Hamiltonian measured in a QNode. Note that this is equivalent to what can be
+  done with `qml.ExpvalCost`, but this is the preferred method because `ExpvalCost` is deprecated.
+  [(#)]()
```
Comment (timmysilv) [inline doc/releases/changelog-dev.md:223]:
```suggestion
  [(#4896)](https://github.com/PennyLaneAI/pennylane/pull/4896)
```

File: pennylane/optimize/shot_adaptive.py [hunk 2]
```diff
@@ -267,6 +267,78 @@ def cost(*args, **kwargs):
...
+            **kwargs: Keyword arguments to the QNode
+
+        Returns:
+            array[float]: the single-shot gradients of the Hamiltonian expectation value
+        """
+        qnode.construct(args, kwargs)
+        base_func = qnode.func
+        tape = qnode.tape
+        [expval] = tape.measurements
+        coeffs, observables = expval.obs.terms()
```
Comment (mudit2812) [inline pennylane/optimize/shot_adaptive.py:292]:
```suggestion
        coeffs, observables = expval.obs.terms() if instance(expval.obs, qml.Hamiltonian) else (1.0, expval.obs)

```
If we're exposing this function as a public method then it's probably worth casting the observable as a `Hamiltonian` first.

**Review by mudit2812 — APPROVED** [review]
Looks good to me! I left a fwe non-blocking suggestions, _mostly_ take it or leave it

File: pennylane/optimize/shot_adaptive.py [hunk 3]
```diff
@@ -267,6 +267,80 @@ def cost(*args, **kwargs):
...
+            if s == 0:
+                continue
+
+            def func(*qnode_args, **qnode_kwargs):
+                qs = qml.tape.make_qscript(base_func)(*qnode_args, **qnode_kwargs)
+                for op in qs.operations:
+                    qml.apply(op)
+                return qml.expval(o)  # pylint:disable=cell-var-from-loop
+
+            qnode.func = func
```
Comment (albi3ro) [inline pennylane/optimize/shot_adaptive.py:317]: Maybe add a copy so we don't have to mutate the `qnode`?

File: pennylane/optimize/shot_adaptive.py [hunk 4]
```diff
@@ -338,6 +410,20 @@ def _single_shot_qnode_gradients(self, qnode, args, kwargs):
         if self.lipschitz is None:
             self.check_learning_rate(1)
 
+        qnode.construct(args, kwargs)
+        [expval] = qnode.tape.measurements
+        if self.term_sampling == "weighted_random_sampling":
+            if isinstance(expval.obs, qml.Hamiltonian):
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/optimize/shot_adaptive.py:418]: > On the other hand, we can change the default value to None since that's what it has effectively been for QNodes. This would be "breaking" for ExpvalCost users, but surely there are none left given the amount of time the deprecation warning has been present for.

This works for me, i.e.,
- `term_sampling=None` by default
- adding a breaking change in the changelog
- Can we also additionally raise a warning in `ShotAdaptiveOptimizer` if an `ExpvalCost` is used, recommending users switch to using `qml.expval(H)` as well as setting `term_sampling='weighted_random_sampling'` - I can help with the warning message here.

[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/optimize/shot_adaptive.py:418]: Can we also update the docstrings, e.g., line 32, 39, 50, to mention these updates.

[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/optimize/shot_adaptive.py:419]: I'd rather not raise warnings in the code, because we only see the type in `opt.step` which might be called many, many times by users.

we already have a [warning in the ExpvalCost doc](https://docs.pennylane.ai/en/stable/code/api/pennylane.ExpvalCost.html), and we [raise a warning](https://github.com/PennyLaneAI/pennylane/blob/b1f68fb9d800aa83a8d81134127a7d6cd7d6644b/pennylane/vqe/vqe.py#L155-L160) when someone creates an `ExpvalCost` instance.

I'll add a similar warning to the `ShotAdaptiveOptimizer` docs

File: pennylane/optimize/shot_adaptive.py [hunk 5]
```diff
@@ -29,10 +30,22 @@ class ShotAdaptiveOptimizer(GradientDescentOptimizer):
...
+        pass Hamiltonians to the :func:`~pennylane.expval` function inside QNodes.
+
+        .. code-block:: python
+
+            @qml.qnode(dev)
+            def ansatz(params):
+                some_qfunc(params)
+                return qml.expval(Hamiltonian)
+
+        In order to optimize the Hamiltonian evaluation taking into account commuting terms, use the ``grouping_type`` keyword in :class:`~.Hamiltonian`.
```
Comment (trbromley) [inline pennylane/optimize/shot_adaptive.py:48]:
```suggestion
        In order to optimize the Hamiltonian evaluation taking into account commuting terms, use the ``grouping_type`` keyword in :class:`~.Hamiltonian`.

        You should also set ``term_sampling='weighted_random_sampling'`` to continue using weighted
        random sampling in ``ShotAdaptiveOptimizer``.
```

File: pennylane/optimize/shot_adaptive.py [hunk 6]
```diff
@@ -267,6 +280,80 @@ def cost(*args, **kwargs):
...
+            for i in argnums:
+                j = qml.jacobian(cost, argnum=i)(*args, **kwargs, shots=new_shots)
+
+                if s == 1:
+                    j = np.expand_dims(j, 0)
+                # Divide each term by the probability per shot. This is
+                # because we are sampling one at a time.
+                jacs.append(c * j / p)
+
+            grads.append(jacs)
```
Comment (trbromley) [inline pennylane/optimize/shot_adaptive.py:353]: @timmysilv please could you mention why it does not make sense to use a transform here to distribute shots?

**Review by trbromley — COMMENTED** [review]
Thanks! No blockers from my side.

**Comment by timmysilv** [comment]
did 2 things on this today:
1. confirmed that instead of doing `res0, res1 = qml.jacobian(cost, argnum=0), qml.jacobian(cost, argnum=1)`, we can just do `res0, res1 = qml.jacobian(cost, argnum=[0, 1])`. So, I  removed the unnecessary for-loop + additional jacobian call, and un-indented all the code which makes the diff ugly.
2. chatted with Romain for some time about trying to make this a transform. The problem is that we need to apply post-processing to the jacobians, not to the execution results. This makes it quite difficult (Romain had tried with other things like `classical_fisher` but found it was not worth the effort), so I'm abandoning my efforts. It _can_ be done if we apply param-shift manually and make it all work, but it's rather overkill for some private code.

File: pennylane/optimize/shot_adaptive.py [hunk 7]
```diff
@@ -29,10 +30,25 @@ class ShotAdaptiveOptimizer(GradientDescentOptimizer):
...
+        pass Hamiltonians to the :func:`~pennylane.expval` function inside QNodes.
+
+        .. code-block:: python
+
+            @qml.qnode(dev)
+            def ansatz(params):
+                some_qfunc(params)
+                return qml.expval(Hamiltonian)
+
+        In order to optimize the Hamiltonian evaluation taking into account commuting terms, use the ``grouping_type`` keyword in :class:`~.Hamiltonian`.
```
Comment (albi3ro) [inline pennylane/optimize/shot_adaptive.py:48]: Does the grouping effect the shot adaptive optimizer at all?

**Review by albi3ro — COMMENTED** [review]
So for:
```
>>> coeffs = [2, 4, -1, 5, 2]
>>> obs = [
...   qml.PauliX(1),
...   qml.PauliZ(1),
...   qml.PauliX(0) @ qml.PauliX(1),
...   qml.PauliY(0) @ qml.PauliY(1),
...   qml.PauliZ(0) @ qml.PauliZ(1)
... ]
>>> H = qml.Hamiltonian(coeffs, obs)
>>> dev = qml.device("default.qubit", wires=2, shots=50)
>>> @qml.qnode(dev)
>>> def cost(weights):
...     qml.StronglyEntanglingLayers(weights, wires=range(2))
...     return qml.expval(H)

>>> dev_analytic = qml.device("default.qubit", wires=2)
>>> @qml.qnode(dev_analytic)
>>> def cost2(weights):
...     qml.StronglyEntanglingLayers(weights, wires=range(2))
...     return qml.expval(H)

shape = qml.templates.StronglyEntanglingLayers.shape(n_layers=2, n_wires=2)
params = pnp.random.random(shape)
opt = qml.ShotAdaptiveOptimizer(min_shots=10, term_sampling="weighted_random_sampling")
costs = []
with dev.tracker:
    for i in range(50):
        params = opt.step(cost, params)
        costs.append(cost2(params))
```

I get:
Screenshot 2023-12-04 at 2 29 13 PM


Quite a number of the expectation values are done with a single shot. This makes me wonder about the `min_shots` keyword argument, since in this case `min_shots=10`.

Without weighted random sampling, the minimum number of shots used is 10.

File: pennylane/optimize/shot_adaptive.py (hunk 7 above)
Comment (timmysilv) [reply pennylane/optimize/shot_adaptive.py:48]: well now that you mention it... it kinda breaks it actually.
- on master with ExpvalCost, it only uses the shots associated with the first term/coeff, scales by that coeff, then raises some mysterious warning
- on this branch with ExpvalCost, autograd fails outright when trying to reshape the resulting jacobian
- with QNode (both versions), the grouping has no effect

this seems like a whole mess - but I'll remove the suggestion to set `grouping_type` for this PR

**Comment by timmysilv** [comment]
WRS does do it with single-shot expectations. couldn't tell you why, but it appears to be on purpose.

File: tests/optimize/test_optimize_shot_adaptive.py [hunk 8]
```diff
@@ -574,6 +574,152 @@ def test_single_shots(self, mocker):
         assert grads[0].shape == (10, *weights.shape)
 
 
+class TestQNodeWeightedRandomSampling:
+    """Tests for weighted random Hamiltonian term sampling"""
+
+    def test_wrs_expval_cost(self, mocker):
```
Comment (mudit2812) [inline tests/optimize/test_optimize_shot_adaptive.py:580]:
```suggestion
    def test_wrs_qnode(self, mocker):

```

**Review by lillian542 — COMMENTED** [review]
The behaviour with the new recommendation seems to differ a bit from the old one. Specifically, using a `qnode` instead of `ExpvalCost`, there are fewer shots per execution, and as a result, the optimization is noisier and requires more reps. Not sure if we want to change the behaviour or possibly just mention it in the docs. For the example in the docstring:

The old method seems to be decently optimized after about 20 reps (corresponding to about 17k total shots used). The x-axis is calls of `opt.step`, y-axis is `cost`:
image

The new method needs closer to 100 reps (around 16k total shots)
image

So they are both working, and ultimately use the same number of shots, but right now if you switch out the `ExpvalCost` for a `qnode` in code that is already working, your optimization is likely to fail because you aren't doing enough reps.

**Comment by josh146** [comment]
> So they are both working, and ultimately use the same number of shots, but right now if you switch out the ExpvalCost for a qnode in code that is already working, your optimization is likely to fail because you aren't doing enough reps.

Ah this sounds like a potential breaking change 🤔 @lillian542 do you have a rough idea why this is the case? Is it an implementation detail in the PR, or something more fundamental that we can't do without `ExpvalCost`?

File: pennylane/optimize/shot_adaptive.py [hunk 9]
```diff
@@ -63,8 +76,8 @@ class ShotAdaptiveOptimizer(GradientDescentOptimizer):
 
     **Example**
 
-    For VQE/VQE-like problems, the objective function for the optimizer can be
-    realized as an :class:`~.ExpvalCost` object, constructed using a :class:`~.Hamiltonian`.
+    For VQE/VQE-like problems, the objective function for the optimizer can be realized
+    as a :class:`~.QNode` object measuring the expectation of a :class:`~.Hamiltonian`.
 
     >>> coeffs = [2, 4, -1, 5, 2]
```
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/optimize/shot_adaptive.py:82]: I had a similar issue but I think our docs still recommend using `from pennylane import numpy as np` in almost all places. The shift away from it is only internal for now

**Comment by lillian542** [comment]
>Ah this sounds like a potential breaking change 🤔 @lillian542 do you have a rough idea why this is the case? Is it an implementation detail in the PR, or something more fundamental that we can't do without ExpvalCost?

@josh146 - actually @timmysilv fixed this in the most recent commit (before my comment), I was having a problem when pulling from GitHub and hadn't noticed, so I was trying out an out-of-date version of this PR. I think he's doing a few more sanity checks to be super sure, but I don't believe there's any issue here. Sorry for the confusion!

**Comment by timmysilv** [comment]
@josh146 - chatted with @lillian542 offline, her local git didn't have the latest fix I pushed so they should be fine. I re-ran the example (with only 30 steps, but same initial params for both QNode and ExpvalCost) a few times, and the results seem fairly consistent. Here are a few comparisons to highlight the chaos from this Optimizer:
fig
fig
fig

**Review by lillian542 — APPROVED** [review]
Looks like reasonable results all around to me!

File: pennylane/optimize/shot_adaptive.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/optimize/shot_adaptive.py:83]: I added the import statement to avoid confusion for devs and users alike

File: pennylane/optimize/shot_adaptive.py [hunk 10]
```diff
@@ -76,7 +89,10 @@ class ShotAdaptiveOptimizer(GradientDescentOptimizer):
...
     >>> H = qml.Hamiltonian(coeffs, obs)
     >>> dev = qml.device("default.qubit", wires=2, shots=100)
-    >>> cost = qml.ExpvalCost(qml.templates.StronglyEntanglingLayers, H, dev)
+    >>> @qml.qnode(dev)
+    >>> def cost(weights):
+    ...     qml.StronglyEntanglingLayers(weights, wires=range(2))
+    ...     return qml.expval(H)
 
     Once constructed, the cost function can be passed directly to the
     optimizer's ``step`` method. The attributes ``opt.shots_used`` and
```
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/optimize/shot_adaptive.py:99]: yes! the default has changed, so this should be updated.
